satellite/metabase/rangedloop/service_test.go

```diff
@@ -117,3 +119,40 @@ func TestLoopDuration(t *testing.T) {
 		require.Equal(t, expectedSleep, observerDuration.Duration.Round(sleepIncrement))
 	}
 }
+
+func TestLoopCancellation(t *testing.T) {
+	parallelism := 2
+	batchSize := 1
+	ctx, cancel := context.WithCancel(testcontext.New(t))
+
+	observers := []rangedloop.Observer{
+		&rangedlooptest.CountObserver{},
+	}
+
+	loopService := rangedloop.NewService(
+		zaptest.NewLogger(t),
+		rangedloop.Config{
+			BatchSize:   batchSize,
+			Parallelism: parallelism,
+		},
+		&rangedlooptest.InfiniteSegmentProvider{},
+		observers,
+	)
+
+	group := errs2.Group{}
+	group.Go(func() error {
+		_, err := loopService.RunOnce(ctx)
+		return err
+	})
+
+	group.Go(func() error {
+		time.Sleep(time.Millisecond * 100)
```

Comment:
I wonder if cancelling from an observer wouldn't be a more robust way of testing this since it would strongly guarantee that the processing loop has been entered at the time of cancellation.

Something like:

```go

observer := &rangedlooptest.CallbackObserver{
     OnProcess: func(ctx context.Context, segments []segmentloop.Segment) error {
        cancel()
     },
}

Reply:
Sounds good. Done.

Reply:
I'm not sure if i'm seeing this change anywhere